build.rs

```diff
@@ -64,159 +64,488 @@
+        asm_args.push(String::from("crypto/libring-asm.Windows.vcxproj"));
+        run_command_with_args(&msbuild, &asm_args);
+    } else {
+        // TODO: rename msvc-ring-asm-i686.lib to msvc-ring-asm-x86.lib
+        let pregenerated_lib_name = format!("msvc-ring-asm-{}.lib",
+                                            target_info.target_arch());
+        let mut pregenerated_lib = PathBuf::from("pregenerated");
+        pregenerated_lib.push(pregenerated_lib_name);
 
-        let mut core_args = args.clone();
-        core_args.push(String::from("crypto/libring.Windows.vcxproj"));
-        run_command_with_args(&msbuild, &core_args);
+        let ring_asm_lib_name = String::from("ring-asm.lib");
+        let mut ring_asm_lib = lib_path.clone();
+        ring_asm_lib.push(&ring_asm_lib_name);
+        println!("{:?} -> {:?}", &pregenerated_lib, &ring_asm_lib);
 
-        let mut test_args = args.clone();
-        test_args.push(String::from("crypto/libring-test.Windows.vcxproj"));
-        run_command_with_args(&msbuild, &test_args);
+        std::fs::create_dir_all(&lib_path).unwrap();
+        let _ = std::fs::copy(&pregenerated_lib, &ring_asm_lib).unwrap();
     }
+    println!("cargo:rustc-link-lib=static=ring-asm");
+
+    let mut core_args = args.clone();
+    core_args.push(String::from("crypto/libring.Windows.vcxproj"));
+    run_command_with_args(&msbuild, &core_args);
 
-    println!("cargo:rustc-link-search=native={}", lib_path.to_str().unwrap());
-    println!("cargo:rustc-link-lib=static={}-core", LIB_NAME);
+    let mut test_args = args.clone();
+    test_args.push(String::from("crypto/libring-test.Windows.vcxproj"));
+    run_command_with_args(&msbuild, &test_args);
 
+    println!("cargo:rustc-link-lib=static=ring-core");
     // XXX: Ideally, this would only happen for `cargo test`, but we don't know
     // how to do that yet.
-    println!("cargo:rustc-link-lib=static={}-test", LIB_NAME);
-    if !use_msbuild {
-        println!("cargo:rustc-flags=-l dylib=stdc++");
-    }
-
-    Ok(())
+    println!("cargo:rustc-link-lib=static=ring-test");
 }
 
 fn run_command_with_args<S>(command_name: S, args: &Vec<String>)
-                            where S: AsRef<std::ffi::OsStr> + Copy {
+    where S: AsRef<std::ffi::OsStr> + Copy
+{
     if !std::process::Command::new(command_name)
-            .args(&args)
-            .status()
-            .unwrap_or_else(|e| { panic!("failed to execute {}: {}",
-                            command_name.as_ref().to_str().unwrap(), e); })
-            .success() {
+        .args(&args)
+        .status()
+        .unwrap_or_else(|e| {
+            panic!("failed to execute {}: {}",
+                   command_name.as_ref().to_str().unwrap(),
+                   e);
+        })
+        .success() {
         panic!("{} execution failed", command_name.as_ref().to_str().unwrap());
     }
 }
+
+fn compile(file: &str, target_info: &TargetInfo, mut out_dir: PathBuf)
```

Comment:
My understanding is that the gcc crate automatically parallelizes the build if you pass in multiple files, and so did the `make`-based approach.  So, we need to decide whether we want to use the gcc crate's parallelism support, or whether we're going to (perhaps in a separate PR, and perhaps not necessarily you writing it) do the parallelism ourselves.

The benefit of doing the parallelism ourselves is that we can do it for  the `*.pl` -> `*.S` and `.S` -> `.o` steps too. The downside to doing the parallelism ourselves is that it is more work than relying on the gcc crate's built-in stuff.

WDYT?